explorer/testdata/linked_list/typed_linked_list.carbon

```diff
@@ -0,0 +1,139 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// AUTOUPDATE
+// RUN: %{explorer-run}
+// RUN: %{explorer-run-trace}
+// CHECK:STDOUT: String array length is: 7
+// CHECK:STDOUT: This is a String Entry 4
+// CHECK:STDOUT: Value 1337
+// CHECK:STDOUT: result: 0
+package ExplorerTest api;
+
+class Node(T:! Type) {
+  fn Create(value: T)-> Node(T) {
+    return {
+      .value = Optional(T).Create(value),
+      .next = Optional(Node(T)*).CreateEmpty(),
+      .prev = Optional(Node(T)*).CreateEmpty()
+    };
+  }
+  fn set_next[addr me:Self*](n: Optional(Node(T)*)) {
+    (*me).next = n;
+  }
+  fn set_prev[addr me:Self*](n: Optional(Node(T)*)) {
+    (*me).prev = n;
+  }
+  var value: Optional(T);
+  var next: Optional(Node(T)*);
+  var prev: Optional(Node(T)*);
+}
+
+class LinkedList(T:! Type) {
+  fn Create() -> LinkedList(T) {
+    return {
+      .head = Optional(Node(T)*).CreateEmpty(),    
+      .tail = Optional(Node(T)*).CreateEmpty(),
+      .len = 0
+    };
+  }
+  fn PushBack[addr me: Self*](value:T) {
+    (*me).len = (*me).len + 1;
+    if(not (*me).head.HasValue()) {
+      (*me).head = Optional(Node(T)*).Create(heap.New(Node(T).Create(value)));
+      (*me).tail = (*me).head;
+      return;
+    }
+    var last: Optional(Node(T)*) = (*me).tail;
+    var last_value:Node(T)* = last.Get(); 
+    var v_wrapped:Optional(Node(T)*) = Optional(Node(T)*).Create(heap.New(Node(T).Create(value)));
+    (*last_value).set_next(v_wrapped);
+    (*v_wrapped.Get()).set_prev(last);
+    (*me).tail = v_wrapped;
+  }
+  fn PushFront[addr me: Self*](value:T) {
+    (*me).len = (*me).len + 1;
+    if(not (*me).head.HasValue()) {
+      (*me).head = Optional(Node(T)*).Create(heap.New(Node(T).Create(value)));
+      (*me).tail = (*me).head;
+      return;
+    }
+    var v_wrapped:Optional(Node(T)*) = Optional(Node(T)*).Create(heap.New(Node(T).Create(value)));
+    var current_head: Optional(Node(T)*) = (*me).head;
+    var current_head_value: Node(T)* = current_head.Get();
+    (*v_wrapped.Get()).set_next(current_head);
+    (*current_head_value).set_prev(v_wrapped);
+    (*me).head = v_wrapped;
+  }
+  fn GetByIndex[me: Self](target_index: i32) -> Optional(T) {
```

Comment:
This is a pretty unusual operation for a linked list to have. What's the rationale for including it?

Reply:
It seamed the most obvious way for me to access a indexed member within the list which was my target, im not entirely sure why this would be **so** unusual, since i think the other way would be to search for a member within the list, which wasn't my target here.

Reply:
> im not entirely sure why this would be so unusual

It's unusual because traversing through a linked list to a particular node is very inefficient, so if your use case needs to do that often enough that you want it to be a member function, that usually means you should use a different data structure.

I think it makes sense for this test to include code that traverses through the linked list, but I think it will be clearer and less surprising if that code is in `Main`, or maybe in a separate helper function that's not a member of `LinkedList`.

Reply:
Yes im aware its ineffecient when used on a larger dataset and in those cases a different datastructure makes more sense, i totally agree! But in this case it was more a utility to solve the problem even though its a linked list.
But either way i do not see a issue with outsourcing the function into a helper and adding a Comment!

Reply:
Addressed in: https://github.com/carbon-language/carbon-lang/pull/2366/commits/48f78338b7202eb358606acd9c420d93600f4782